clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/HelixScatterGatherRoutingStrategy.java

```diff
@@ -0,0 +1,63 @@
+package com.linkedin.venice.fastclient.meta;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+
+/**
+ * This strategy will assign an ordering of helix groups via round-robin for even distribution. The fanout in multi-get
+ * will be performed using instances belonging to the assigned groups. If no instance belonging to the selected group is
+ * found for a given partition, the instance with the next group in the assigned ordering will be used
+ */
+public class HelixScatterGatherRoutingStrategy implements ClientRoutingStrategy {
+  private Map<String, Integer> helixGroupInfo;
+
+  public HelixScatterGatherRoutingStrategy(Map<String, Integer> helixGroupInfo) {
+    this.helixGroupInfo = helixGroupInfo;
+  }
+
+  @Override
+  public List<String> getReplicas(long requestId, List<String> replicas, int requiredReplicaCount) {
+    if (replicas.isEmpty()) {
+      return Collections.emptyList();
+    }
+    Map<Integer, List<String>> groupToFilteredReplicas = new HashMap<>();
+    for (String replica: replicas) {
+      groupToFilteredReplicas.computeIfAbsent(helixGroupInfo.get(replica), k -> new ArrayList<>()).add(replica);
+    }
+    List<Integer> shuffledGroupIds = shuffleGroupIds(groupToFilteredReplicas.keySet());
```

Comment:
I'm thinking is there a better way to do this because unavailable replica for a given group should be an edge case and creating a new list of shuffled groupId on every request sounds expensive for edge case handling. What if we simply try a random group if no replica exists in the selected group?

Reply:
Yeah this makes sense, since it's an edge case we don't care too much about getting even distribution past the first

Reply:
addressed in latest commit